### ISSUE TITLE
Lazyloading of state modules happening in wrong order

### ISSUE THREAD

**Issue by wt** [saltstack/salt#26115]
salt-call --versions:

               Salt: 2015.5.3
             Python: 2.6.9 (unknown, Apr  1 2015, 18:16:00)
             Jinja2: 2.7.2
           M2Crypto: 0.21.1
     msgpack-python: 0.4.6
       msgpack-pure: Not Installed
           pycrypto: 2.6.1
            libnacl: Not Installed
             PyYAML: 3.10
              ioflo: Not Installed
              PyZMQ: 14.3.1
               RAET: Not Installed
                ZMQ: 3.2.5
               Mako: Not Installed
            Tornado: Not Installed

I believe that I have a case where the ordering of states is not working properly.

I have a file root with the following files in it:

postfix.sls:

    postfix:
      service.running

python-pip.sls:

    python_pip_pkg:
      pkg.latest:
        - pkgs:
          - python26-pip
          - python27-pip

docker-py.sls:

    include:
      - base: python-pip
    
    docker-py:
      pip.installed:
        - name: docker-py>=1.2.2
        - bin_env: /usr/bin/pip-2.7
        - watch_in:
          - service: postfix

The 
Here a result of trying to use them before installing pip:

    # salt-call --local --file-root /root/blah state.apply postfix,docker-py
    local:
        - The function "state.highstate" is running as PID 13502 and was started at 2015, Aug 07 20:49:59.225647 with jid 20150807204959225647
    # salt-call --local --file-root /root/blah state.apply postfix,docker-py 2>&1 | sed 's/^/    /'
    [INFO    ] Loading fresh modules for state activity
    [INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://postfix.sls'
    [INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://docker-py.sls'
    [INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://python-pip.sls'
    /usr/lib64/python2.6/dist-packages/Crypto/Util/number.py:57: PowmInsecureWarning: Not using mpz_powm_sec.  You should rebuild using libgmp >= 5 to avoid timing attack vulnerability.
      _warn("Not using mpz_powm_sec.  You should rebuild using libgmp >= 5 to avoid timing attack vulnerability.", PowmInsecureWarning)
    [ERROR   ] An un-handled exception was caught by salt's global exception handler:
    KeyError: 'pip.installed'
    Traceback (most recent call last):
      File "/usr/bin/salt-call", line 11, in <module>
        salt_call()
      File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 227, in salt_call
        client.run()
      File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 69, in run
        caller.run()
      File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 236, in run
        ret = self.call()
      File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 138, in call
        ret['return'] = func(*args, **kwargs)
      File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 296, in apply_
        return sls(mods, **kwargs)
      File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 681, in sls
        ret = st_.state.call_high(high_)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 2067, in call_high
        ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 1623, in call_chunks
        running = self.call_chunk(low, running, chunks)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 1858, in call_chunk
        running = self.call_chunk(chunk, running, chunks)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 1769, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 612, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
        raise KeyError(key)
    KeyError: 'pip.installed'
    Traceback (most recent call last):
      File "/usr/bin/salt-call", line 11, in <module>
        salt_call()
      File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 227, in salt_call
        client.run()
      File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 69, in run
        caller.run()
      File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 236, in run
        ret = self.call()
      File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 138, in call
        ret['return'] = func(*args, **kwargs)
      File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 296, in apply_
        return sls(mods, **kwargs)
      File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 681, in sls
        ret = st_.state.call_high(high_)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 2067, in call_high
        ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 1623, in call_chunks
        running = self.call_chunk(low, running, chunks)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 1858, in call_chunk
        running = self.call_chunk(chunk, running, chunks)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 1769, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 612, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
        raise KeyError(key)
    KeyError: 'pip.installed'

Running the previous command with debug logging appears to indicate that the pip module is LazyLoaded just before the crash and before it ever tries applying any of the states. I expected the python-pip state file to be applied first since it is included from the docker-py which actually happens if I remove the watch_in requisite from the docker-py. I actually think this is an important piece of info for this bug.

I also tried other cases.

If I install pip before running the command, it will work.

If I swap the postfix and docker-py states on the commandline before I install pip, it also works:

    # salt-call --local --file-root /root/blah state.apply docker-py,postfix
    [INFO    ] Loading fresh modules for state activity
    [INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://docker-py.sls'
    [INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://python-pip.sls'
    [INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://postfix.sls'
    [INFO    ] Running state [python_pip_pkg] at time 20:52:17.637019
    [INFO    ] Executing state pkg.latest for python_pip_pkg
    [INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
    [INFO    ] Executing command ['yum', '-q', 'clean', 'expire-cache'] in directory '/root'
    [INFO    ] Executing command ['yum', '-q', 'check-update'] in directory '/root'
    [INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}'   --pkgnarrow=available python26-pip python27-pip" in directory '/root'
    [INFO    ] Executing command 'yum -y     install python26-pip python27-pip' in directory '/root'
    [INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
    [INFO    ] Installed Packages:
    python26-pip changed from absent to 6.1.1-1.20.amzn1
    python27-pip changed from absent to 6.1.1-1.20.amzn1
    
    [INFO    ] Loading fresh modules for state activity
    [INFO    ] Completed state [python_pip_pkg] at time 20:52:31.346377
    [INFO    ] Running state [docker-py>=1.2.2] at time 20:52:31.483259
    [INFO    ] Executing state pip.installed for docker-py>=1.2.2
    [INFO    ] Executing command '/usr/bin/pip-2.7 freeze' in directory '/root'
    [INFO    ] Executing command '/usr/bin/pip-2.7 install' in directory '/root'
    [INFO    ] Python package docker-py>=1.2.2 was already installed
    All packages were successfully installed
    [INFO    ] Completed state [docker-py>=1.2.2] at time 20:52:31.967712
    [INFO    ] Running state [postfix] at time 20:52:31.969605
    [INFO    ] Executing state service.running for postfix
    [INFO    ] Executing command '/sbin/service postfix status' in directory '/root'
    [INFO    ] The service postfix is already running
    [INFO    ] Completed state [postfix] at time 20:52:31.988342
    local:
        ----------
        pip_|-docker-py_|-docker-py>=1.2.2_|-installed:
            ----------
            __run_num__:
                1
            changes:
                ----------
            comment:
                Python package docker-py>=1.2.2 was already installed
                All packages were successfully installed
            duration:
                484.453
            name:
                docker-py>=1.2.2
            result:
                True
            start_time:
                20:52:31.483259
        pkg_|-python_pip_pkg_|-python_pip_pkg_|-latest:
            ----------
            __run_num__:
                0
            changes:
                ----------
                python26-pip:
                    ----------
                    new:
                        6.1.1-1.20.amzn1
                    old:
                python27-pip:
                    ----------
                    new:
                        6.1.1-1.20.amzn1
                    old:
            comment:
                The following packages were successfully installed/upgraded: python26-pip, python27-pip
            duration:
                13709.358
            name:
                python_pip_pkg
            result:
                True
            start_time:
                20:52:17.637019
        service_|-postfix_|-postfix_|-running:
            ----------
            __run_num__:
                2
            changes:
                ----------
            comment:
                The service postfix is already running
            duration:
                18.737
            name:
                postfix
            result:
                True
            start_time:
                20:52:31.969605

If I explicitly prepend the applied states with the "python-pip" state file, it works:

    # salt-call --local --file-root /root/blah state.apply python-pip,postfix,docker-py
    ...

If I remove the "watch_in" requisite from the docker-py state in the docker-py.sls file, it will also work. However, I want to restart the postfix service whenever docker-py is changed, so I actually do want that requisite.

Just so you know, the actual daemon that I am dealing with is one in use at my company. I am just using postfix so that the examples can work more easily for others.

**Comment by wt** [comment]
I am pretty sure this behavior is different between 2015.5.3 and 2015.5.2.

**Comment by anlutro** [comment]
As a workaround, if you don't want to put a `require: - pkg: pip` in all of your pip-related states, you can add `- order: 0` to the pip state to ensure that it runs first. I do this for pip, python-mysql and other python packages that salt utilize.

**Comment by silenius** [comment]
Hello,

I don't know if it's related but I have a similar error:

```
[DEBUG   ] Could not LazyLoad pip.installed
[WARNING ] The minion function caused an exception
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/salt/minion.py", line 1161, in _thread_return
    return_data = func(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/salt/modules/state.py", line 681, in sls
    ret = st_.state.call_high(high_)
  File "/usr/local/lib/python2.7/site-packages/salt/state.py", line 2067, in call_high
    ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
  File "/usr/local/lib/python2.7/site-packages/salt/state.py", line 1623, in call_chunks
    running = self.call_chunk(low, running, chunks)
  File "/usr/local/lib/python2.7/site-packages/salt/state.py", line 1769, in call_chunk
    self._mod_init(low)
  File "/usr/local/lib/python2.7/site-packages/salt/state.py", line 612, in _mod_init
    self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
  File "/usr/local/lib/python2.7/site-packages/salt/utils/lazy.py", line 90, in __getitem__
    raise KeyError(key)
KeyError: 'pip.installed'
```

This is with the following state file:

```salt
{{ venv }}:
    virtualenv.managed:
        - name: {{ cfg.full_path }}
        {% if cfg.python_version == 3.4 %}
        - venv_bin: /usr/local/bin/pyvenv-3.4
        {% elif cfg.python_version == 2.7 %}
        - venv_bin: /usr/local/bin/virtualenv-2.7
        {% endif %}
        - system_site_packages: False
        - user: {{ cfg.user }}
        - clear: True
        - require:
            - file: {{ cfg.full_path }}
            - user: {{ cfg.user }}
            - pkg: {{ salt['pillar.get']('postgresql:lookup:pkg_client') }}
            {% if cfg.python_version == 3.4 %}
            - pkg: python34
            {% elif cfg.python_version == 2.7 %}
            - pkg: python27
            - pkg: py27-virtualenv
            {% endif %}

{{ venv }}_upgrade_pip:
    cmd.run:
        - name: {{ cfg.full_path }}/bin/pip --no-cache-dir install -U pip
        - user: {{ cfg.user }}
        - group: {{ cfg.user }}
        - reload_modules: True
        - require:
            - virtualenv: {{ venv }}

{{ venv }}_requirements:
    pip.installed:
        - name: ''
        - bin_env: {{ cfg.full_path }}
        - cwd: {{ cfg.full_path }}
        - requirements: salt://venvs/webapps/files/{{ venv }}-requirements.txt
        - user: {{ cfg.user }}
        - use_wheel: False
        - use_vt: True
        - reload_modules: True
        - env_vars:
            PIP_CACHE_DIR: /tmp
        - watch:
            - cmd: {{ venv }}_upgrade_pip
```

**Comment by silenius** [comment]
I think that the problem is with the following code (in states/pip_state.py):

```python
try:
    import pip
    HAS_PIP = True
except ImportError:
    HAS_PIP = False

if HAS_PIP is True:
    try:
        import pip.req
    except ImportError:
        HAS_PIP = False
        # Remove references to the loaded pip module above so reloading works
        import sys
        del pip
        if 'pip' in sys.modules:
            del sys.modules['pip']
# pylint: enable=import-error

logger = logging.getLogger(__name__)

# Define the module's virtual name
__virtualname__ = 'pip'


def __virtual__():
    '''
    Only load if the pip module is available in __salt__
    '''
    if HAS_PIP and 'pip.list' in __salt__:
        return __virtualname__
    return False

```

pip could not be installed globally but locally, once the virtualenv has been created:

```
jcigar@dev:~/ > pip      
zsh: command not found: pip
jcigar@dev:~/ > python
Python 2.7.10 (default, Jul 26 2015, 11:57:29) 
[GCC 4.2.1 Compatible FreeBSD Clang 3.4.1 (tags/RELEASE_34/dot1-final 208032)] on freebsd10
Type "help", "copyright", "credits" or "license" for more information.
>>> import pip
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named pip
>>> 
jcigar@dev:~/ > source /usr/local/venvs/harmonia/bin/activate
(harmonia)jcigar@dev:~/ > pip

Usage:   
  pip <command> [options]

Commands:
  install                     Install packages.
  uninstall                   Uninstall packages.
  freeze                      Output installed packages in requirements format.
  list                        List installed packages.
  show                        Show information about installed packages.
  search                      Search PyPI for packages.
  wheel                       Build wheels from your requirements.
  zip                         DEPRECATED. Zip individual packages.
  unzip                       DEPRECATED. Unzip individual packages.
  bundle                      DEPRECATED. Create pybundles.
  help                        Show help for commands.

General Options:
  -h, --help                  Show help.
  -v, --verbose               Give more output. Option is additive, and can be used up to 3 times.
  -V, --version               Show version and exit.
  -q, --quiet                 Give less output.
  --log-file <path>           Path to a verbose non-appending log, that only logs failures. This log is active by default at /home/jcigar/.pip/pip.log.
  --log <path>                Path to a verbose appending log. This log is inactive by default.
  --proxy <proxy>             Specify a proxy in the form [user:passwd@]proxy.server:port.
  --timeout <sec>             Set the socket timeout (default 15 seconds).
  --exists-action <action>    Default action when a path already exists: (s)witch, (i)gnore, (w)ipe, (b)ackup.
  --cert <path>               Path to alternate CA bundle.
(harmonia)jcigar@dev:~/ > python
Python 2.7.10 (default, Jul 26 2015, 11:57:29) 
[GCC 4.2.1 Compatible FreeBSD Clang 3.4.1 (tags/RELEASE_34/dot1-final 208032)] on freebsd10
Type "help", "copyright", "credits" or "license" for more information.
>>> import pip
>>> import pip.req
>>> 
(harmonia)jcigar@dev:~/ > deactivate 
jcigar@dev:~/ > 
```

**Comment by jfindlay** [comment]
@wt, thanks for the report.  I am curious if #26160 fixes this issue for you.

**Comment by wt** [comment]
I just tried the fix. It does seem to work when I patch my pip_state.py file.

**Comment by wt** [comment]
MAYDAY! We have a similar issue with the git state module.

**Comment by wt** [comment]
Actually, given the fact that I am seeing this with multiple state modules, is there some underlying fix that needs to be made instead of fixing each of the state modules themselves?

**Comment by silenius** [comment]
can you post a stacktrace of, let's say, the issue with the git module ?

**Comment by wt** [comment]
Add /root/blah/metricproxy.sls to the above files with contents:

```
include:
  - base: postfix

metricproxy_git_repo:
  git.latest:
    - name: git@github.com:signalfx/metricproxy.git
    - rev: master
    - target: /root/repos/metricproxy
    - watch_in:
      - service: postfix
```

Now try to apply that state like so:
```
# salt-call --local --file-root /root/blah state.apply postfix,metricproxy
[INFO    ] Loading fresh modules for state activity
[INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://postfix.sls'
[INFO    ] Fetching file from saltenv 'base', ** done ** 'metricproxy.sls'
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
KeyError: 'git.latest'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 227, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 69, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 236, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 296, in apply_
    return sls(mods, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 681, in sls
    ret = st_.state.call_high(high_)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2067, in call_high
    ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1623, in call_chunks
    running = self.call_chunk(low, running, chunks)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1858, in call_chunk
    running = self.call_chunk(chunk, running, chunks)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1769, in call_chunk
    self._mod_init(low)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 612, in _mod_init
    self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
  File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
    raise KeyError(key)
KeyError: 'git.latest'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 227, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 69, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 236, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 296, in apply_
    return sls(mods, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 681, in sls
    ret = st_.state.call_high(high_)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2067, in call_high
    ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1623, in call_chunks
    running = self.call_chunk(low, running, chunks)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1858, in call_chunk
    running = self.call_chunk(chunk, running, chunks)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1769, in call_chunk
    self._mod_init(low)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 612, in _mod_init
    self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
  File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
    raise KeyError(key)
KeyError: 'git.latest'
```

Same game, different name.

**Comment by silenius** [comment]
This may sound a silly question, but: is git installed ? (check with `which git`)

**Comment by sebest** [comment]
Hi,

I had a similar issue, and the problem was that pip was not detected in pip_state.py because of an import error:

Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/states/pip_state.py", line 35, in <module>
    import pip
  File "/usr/lib/python2.7/dist-packages/pip/__init__.py", line 61, in <module>
    from pip.vcs import git, mercurial, subversion, bazaar  # noqa
  File "/usr/lib/python2.7/dist-packages/pip/vcs/mercurial.py", line 9, in <module>
    from pip.download import path_to_url
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 25, in <module>
    from requests.compat import IncompleteRead
ImportError: cannot import name IncompleteRead

the problem seems to be ubuntu specific:
http://stackoverflow.com/questions/27341064/how-do-i-fix-importerror-cannot-import-name-incompleteread

i think that not using the pip provided by the ubuntu should fix the issue

**Comment by sebest** [comment]
i can confirm that removing the package python-pip and installing pip with easy_install pip fixed my issue

**Comment by sebest** [comment]
the real issue seems to be a conflict between the old pip version provided by ubuntu and a newer version of request (maybe installed by salt?)

**Comment by LarsFronius** [comment]
thanks @sebest - also ran into this and an easy_install pip fixed it for me.
Anyone know how we could fix this in the actual ubuntu package of salt? I bet this is something many people will stumble upon.

**Comment by kracekumar** [comment]
+1. I am facing this issue.

**Comment by silenius** [comment]
did you try with:
```salt
- reload_modules: True
```
?

**Comment by jfindlay** [comment]
@sebest, @LarsFronius, can you paste any output or errors or log information related to the conflict between an old pip and requests?  I would rather salt check for pip versions if possible before packaging and depending on a newer version of pip.

**Comment by sebest** [comment]
@jfindlay the error message is the one i pasted in a previous comment with the stackoverflow thread about the changed in the requests API.

**Comment by sebest** [comment]
@jfindlay I think that salt should not depend on OS level libraries and should come in a big package with all its dependencies built-in, that a major issue with languages like python and ruby and that's why many admin tool like that are moving to golang.

Datadog has a blog about packaging a python app as a self contained app with all its dependencies: https://www.datadoghq.com/blog/new-datadog-agent-omnibus-ticket-dependency-hell/

It is using https://github.com/chef/omnibus-software, i guess that Chef did the exact same thing for the exact same reason.

I'm pretty confident that many bugs and regressions in Salt comes from the different versions of python and different versions of its dependencies on the OS.

**Comment by kracekumar** [comment]
+1 for @sebest point about packaging. [Requests ](https://github.com/kennethreitz/requests/tree/master/requests/packages) vendors dependencies for similar reason.

**Comment by Deshke** [comment]
current workaround on 2015.5.3 (Lithium)


```yaml

python-pip:
  pkg.installed
pip install docker-py:
  cmd.run:
  - require:
      - pkg: python-pip

```

**Comment by chmarr** [comment]
Just in case it helps anyone, I managed to track down part of the problem, especially why running "easy_install pip" seems to fix the problem. There's only two things that the easy_install does to the installation... it drops a bunch of entry point programs into /usr/local/bin, which seems to be immaterial, bt it also drops a easy_install.pth into /usr/local/lib/python2.7/dist-packages ... which does a few things:

1. makes sure /usr/lib/python2.7/dist-packages is in sys.path (which it would have been absent easy_install.pth.. something further down the track adds in that path), but...
2. Makes sure that /usr/lib/python2.7/dist-packages appears towards the front of sys.path. This is the clincher. If the "local" version of dist-packages appears first, then salt-* will not be able to find the pip_state/pip modules as above.

This implies that there's some library incompatibility thing going on. examining the two dist-packages directory i see that there are duplicates of the msgpack and requests models, with different version numbers. non-"local" has 0.3.0 and 2.2.1, and "local" has 0.4.6 and 2.7.0.

I'm going to do a bit of fiddling around to see if I can figure out how those additional packages got into there.. be it my own machinations or part of the bootstrap script.

**Comment by rallytime** [comment]
Hi All - This looks like the same regression/stacktrace as reported in #27619, #26882, #26454, #27016, and #26280 and should be fixed 2015.5.6 and 2015.8.0. Can anyone here test and confirm this either at one of those tags or at the HEAD of the 2015.8 or 2015.5 branch?

The backstory here is that this bug was fixed on the 2015.8 brach with #27681, which I back-ported to 2015.5 with #27681.

**Comment by rallytime** [comment]
Update: It appears that I misjudged when my back-porting PR went into the 2015.5 branch and it missed the 2015.5.6 release tag. However, the initial stacktrace reported by @wt both in the pip and git states should be fixed at the HEAD of the 2015.5 branch as well as 2015.8.0.

**Comment by rallytime** [comment]
Since this is fixed at the Head of the 2015.5 branch and will be available in the upcoming 2015.5.7 release (and is fixed in 2015.8.0 and newer), I am going to close this bug. If this pops up again, let us know and we can take another look. Thanks!
